type-metadata/src/lib.rs

```diff
@@ -1,7 +1,346 @@
+	I128,
+	Array(ArrayIdent),
+	Slice(SliceIdent),
+	Tuple(TupleIdent),
+	Option(OptionIdent),
+	Result(ResultIdent),
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct Namespace(Vec<&'static str>);
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct CustomIdent {
+	pub name: &'static str,
+	pub namespace: Namespace,
+	pub type_params: Vec<IdentKind>,
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct ArrayIdent {
+	pub len: u16,
+	pub type_param: Box<IdentKind>,
+}
+impl ArrayIdent {
+	fn new(len: u16, type_param: IdentKind) -> Self {
+		ArrayIdent {
+			len,
+			type_param: Box::new(type_param),
+		}
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct TupleIdent {
+	pub type_params: Option<Vec<IdentKind>>,
+}
+impl TupleIdent {
+	fn new(type_params: Option<Vec<IdentKind>>) -> Self {
+		TupleIdent { type_params }
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct SliceIdent {
+	pub type_param: Box<IdentKind>,
+}
+impl SliceIdent {
+	fn new(type_param: IdentKind) -> Self {
+		SliceIdent {
+			type_param: Box::new(type_param),
+		}
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct OptionIdent {
+	pub type_param: Box<IdentKind>,
+}
+impl OptionIdent {
+	fn new(type_param: IdentKind) -> Self {
+		OptionIdent {
+			type_param: Box::new(type_param),
+		}
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct ResultIdent {
+	pub type_params: (Box<IdentKind>, Box<IdentKind>),
+}
+impl ResultIdent {
+	fn new(type_params: (IdentKind, IdentKind)) -> Self {
+		ResultIdent {
+			type_params: (Box::new(type_params.0), Box::new(type_params.1)),
+		}
+	}
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum TypeDef {
+	Primitive,
+	Struct(StructDef),
+	TupleStruct(TupleStructDef),
+	Enum(EnumDef),
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct Field {
+	pub name: &'static str,
+	pub ident: IdentKind,
+}
+#[derive(PartialEq, Eq, Debug)]
+pub struct StructDef(Vec<Field>);
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct TupleStructDef(Vec<IdentKind>);
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct DataVariant<T> {
+	pub name: &'static str,
+	pub index: u16,
+	pub struct_def: T,
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct NoDataVariant {
+	pub name: &'static str,
+	pub index: u16,
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum VariantKind {
+	NoData(NoDataVariant),
+	Struct(DataVariant<StructDef>),
+	TupleStruct(DataVariant<TupleStructDef>),
+}
+#[derive(PartialEq, Eq, Debug)]
+pub struct EnumDef(Vec<VariantKind>);
+
+pub trait Metadata {
+	fn type_ident() -> IdentKind;
+
+	/// If the current type contains any other types, `type_def` would register their metadata into the given
+	/// `registry`. For instance, `<Option<MyStruct>>::type_def()` would register `MyStruct` metadata. All
+	/// implementation must register these contained types' metadata.
+	fn type_def(registry: &mut Registry) -> TypeDef;
```

Comment:
As far as I understood we can move this into a free function with a simple conversion from `IdentKind` to `TypeDef` since it is always just the same and by moving it out the trait becomes simpler for all implementors.

Reply:
If efficiency is a problem (which I don't believe here) we might want to annotate the free function with an `#[inline(always)]` so that the compiler sees that the `type_ident` is always the same and that the more complex implementation of the generic `type_def` function can be optimized away.

Reply:
`type_def` meant to be implemented

Reply:
`type_def` are supposed to be implemented by all `T: Metadata`, so that with `pub struct A { b: B }`, when `A::type_def()` called, `A` could be guaranteed that `B` knows about it's own `TypeDef`, so that `A` could be able to return complete metadata info(including any related types metadata to fully define `A`).